### DISCUSSION TITLE
asyncio: RuntimeError: Event loop is closed

### DISCUSSION THREAD

**Discussion by mgrandi** [sqlalchemy/sqlalchemy#5994] [Usage Questions] [answered]
**Describe the bug**

i get `RuntimeError: Event loop is closed` after running the below code snippet

**Expected behavior**

not to get the `RuntimeError: Event loop is closed` error message after the program exits

**To Reproduce**
Please try to provide a [Minimal, Complete, and Verifiable](http://stackoverflow.com/help/mcve) example.
See also [Reporting Bugs](https://www.sqlalchemy.org/participate.html#bugs) on the website, and some [example issues](https://github.com/sqlalchemy/sqlalchemy/issues?q=label%3A%22great+mcve%22).

**Please do not use Flask-SQLAlchemy or any other third-party extensions or dependencies in test cases**.  The test case must illustrate the problem **without** using any third party SQLAlchemy extensions.  Otherwise, please report the bug to those projects first.

```py
import asyncio
import logging
import random

from sqlalchemy import Column, Index, Integer, Unicode, PrimaryKeyConstraint
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy_repr import RepresentableBase
from sqlalchemy.engine.url import URL
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.asyncio import AsyncSession

logging.basicConfig(level="DEBUG")
logger = logging.getLogger("main")

CustomDeclarativeBase = declarative_base(name="CustomDeclarativeBase")

class UsersFound(CustomDeclarativeBase):
    __tablename__ = "users_found"

    # primary key column
    user_id = Column(Integer, nullable=False, autoincrement=True)

    user_name = Column(Unicode, nullable=False)

    __table_args__ = (
        PrimaryKeyConstraint("user_id", name="PK-users_found-user_id"),

        Index("IXUQ-users_found-user_name", "user_name", unique=True),

    )

    def __str__(self):
        return f"UsersFound(user_id={self.user_id}, user_name={self.user_name})"


class ScrapeUsers:



    def __init__(self):

        self.config = None
        self.sqla_engine = None
        self.async_sessionmaker = None

    async def run(self):

        url = URL.create(drivername="postgresql+asyncpg",
            username="postgres",
            password="PASSWORDHERE",
            host="localhost",
            port="5432",
            database="urls",
            query=None)
        self.sqla_engine = create_async_engine(url, echo=False, future=True)


        # expire_on_commit=False will prevent attributes from being expired
        # after commit.
        self.async_sessionmaker = sessionmaker(
            bind=self.sqla_engine, expire_on_commit=False, class_=AsyncSession
        )

        async with self.sqla_engine.begin() as conn:
            await conn.run_sync(CustomDeclarativeBase.metadata.create_all)


        async with self.async_sessionmaker() as session:

            # get things from the queue
            async with session.begin():

                name_list_letters = random.sample("abcdefghijklmnopqrstuvwxyz", 10)
                name = "".join(name_list_letters)
                obj = UsersFound(user_name=name)

                logger.info("adding: `%s`", obj)

                session.add(obj)


x = ScrapeUsers()
asyncio.run(x.run(), debug=True)
```

**Error**

```
python C:\Users\mark\Temp\test_asyncio_sqla_bug.py
DEBUG:asyncio:Using proactor: IocpProactor
DEBUG:asyncio:Get address info localhost:5432, type=<SocketKind.SOCK_STREAM: 1>
DEBUG:asyncio:Getting address info localhost:5432, type=<SocketKind.SOCK_STREAM: 1> took 0.000ms: [(<AddressFamily.AF_INET6: 23>, <SocketKind.SOCK_STREAM: 1>, 0, '', ('::1', 5432, 0, 0)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 0, '', ('127.0.0.1', 5432))]
DEBUG:asyncio:<asyncio.TransportSocket fd=636, family=AddressFamily.AF_INET6, type=SocketKind.SOCK_STREAM, proto=0, laddr=('::1', 58256, 0, 0), raddr=('::1', 5432, 0, 0)> connected to localhost:5432: (<_ProactorSocketTransport fd=636 read=<_OverlappedFuture pending overlapped=<pending, 0x2ac308e2af0> cb=[_ProactorReadPipeTransport._loop_reading()] created at c:\python39\lib\asyncio\windows_events.py:465>>, <asyncpg.connect_utils.TLSUpgradeProto object at 0x000002AC30846DC0>)
INFO:main:adding: `UsersFound(user_id=None, user_name=rqdwlvigts)`
DEBUG:asyncio:Close <ProactorEventLoop running=False closed=False debug=True>
Exception ignored in: <function _ProactorBasePipeTransport.__del__ at 0x000002AC2EFC49D0>
Traceback (most recent call last):
  File "c:\python39\lib\asyncio\proactor_events.py", line 116, in __del__
  File "c:\python39\lib\asyncio\proactor_events.py", line 108, in close
  File "c:\python39\lib\asyncio\base_events.py", line 746, in call_soon
  File "c:\python39\lib\asyncio\base_events.py", line 510, in _check_closed
RuntimeError: Event loop is closed
```

**Versions.**
 - OS: windows 10
 - Python: `Python 3.9.1 (tags/v3.9.1:1e5d33e, Dec  7 2020, 17:08:21) [MSC v.1927 64 bit (AMD64)]`
 - SQLAlchemy: `1.4.0b3`
 - Database: postgres 13
 - DBAPI: `asyncpg` 0.22.0

**Additional context**

**Have a nice day!**

**Comment by zzzeek** [comment]
hi -

im on unix so no error is reproduced.  however, it looks like perhaps your engine is getting garbage collected because you have it assigned to an instance variable.   make sure you dispose it within the event loop:

        await self.sqla_engine.dispose()

**Reply by mgrandi** [reply]
thanks! that seemed to be the issue, see the comment thread by CaseIIT

**Reply by alk-sramond** [reply]
FYI i had the same issue on unix (ubuntu 20) (using the sample from the doc) and dispose worked also:

```
async def async_main():
    engine = create_async_engine(
        "mysql+aiomysql://root:toor@172.17.0.3/test",
        echo=False,
    )

    (...)
    await engine.dispose()


asyncio.run(async_main())
```

**Reply by CaselIT** [reply]
ok, I'll remove "on windows" from the title

**Reply by whg517** [reply]
I think this should be written in the document. This operation is not mentioned in the [relevant cases](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#asynchronous-i-o-asyncio).

**Comment by CaselIT** [comment, marked as answer]
@mgrandi I can reproduce the issue, it seems that the windows implementation of asyncio is a bit more finicky compared to the unix one.

That said I don't think sqlalchemy can do anything here. The problem is that the engine is not disposed when the loop exits and it cannot be gc-end since it's still referenced by the instance of ScrapeUsers.
The same error happen using asyncpg directly:

```py
import asyncpg
...
class ScrapeUsers:
    def __init__(self):

        self.conn = None
    async def run(self):
        self.conn = await asyncpg.connect(user="scott", password="tiger", database="test")
        print(await self.conn.fetch('select 1'))

x = ScrapeUsers()
asyncio.run(x.run(), debug=True)
```
results in
```
[<Record ?column?=1>]
Exception ignored in: <function _ProactorBasePipeTransport.__del__ at 0x000002630AC6B550>
Traceback (most recent call last):
  File "C:\...\lib\asyncio\proactor_events.py", line 116, in __del__
  File "C:\...\lib\asyncio\proactor_events.py", line 108, in close
  File "C:\...\lib\asyncio\base_events.py", line 719, in call_soon
  File "C:\...\lib\asyncio\base_events.py", line 508, in _check_closed
RuntimeError: Event loop is closed
```

The same happens with aiomysql btw, so it seems a common thing on windows.

As @zzzeek suggested you can call ` await self.sqla_engine.dispose()` before exiting the `run` function.
Another alternative is to use the `NullPool` with the engine, passing `poolclass=NullPool` to `create_async_engine`

**Reply by mgrandi** [reply]
hmm, is this a bug in python then that i should report? strange that it works on Unix but not windows

but yes, indeed the problem was that i was not disposing of the engine, doing a `await self.sqla_engine.dispose()` fixed it 

However, I also noticed that it was still happening in the codebase i was working in, and i read somewhere that i needed to put an `asyncio.sleep(1)` at the end, and that seemed to fix it. Is that also something I am doing wrong?

**Reply by CaselIT** [reply]
Not sure. I think that the problem is that some object destroyer tries to use async stuff and it fails when asyncio has stopped.

I'm not sure if this is expected of not, it usually always best to cleanup before closing :)

**Reply by mgrandi** [reply]
ok, after doing some googling it seems to be an issue / caveat with aiohttp where you need to sleep to make sure all the connections are closed, see https://docs.aiohttp.org/en/stable/client_advanced.html?highlight=proxy#graceful-shutdown

thanks for your help
